components/content_settings/core/browser/brave_content_settings_policy_provider.cc

```diff
@@ -0,0 +1,144 @@
+#include "components/content_settings/core/browser/content_settings_policy_provider.h"
+#include "components/content_settings/core/browser/content_settings_utils.h"
+#include "components/content_settings/core/browser/website_settings_registry.h"
+#include "components/content_settings/core/common/content_settings_pattern.h"
+#include "components/pref_registry/pref_registry_syncable.h"
+#include "components/prefs/pref_service.h"
+
+namespace {
+
+// Duplicating declaration from
+// components/content_settings/core/browser/content_settings_policy_provider.cc
+struct PrefsForManagedContentSettingsMapEntry {
+  const char* pref_name;
+  ContentSettingsType content_type;
+  ContentSetting setting;
+};
+
+constexpr PrefsForManagedContentSettingsMapEntry
+    kBravePrefsForManagedContentSettingsMap[] = {
+        {kManagedShieldsDisabledForUrls, ContentSettingsType::BRAVE_SHIELDS,
+         CONTENT_SETTING_BLOCK},
+        {kManagedShieldsEnabledForUrls, ContentSettingsType::BRAVE_SHIELDS,
+         CONTENT_SETTING_ALLOW}};
+
+constexpr const char* kBraveManagedPrefs[] = {kManagedShieldsDisabledForUrls,
+                                              kManagedShieldsEnabledForUrls};
+
+}  // namespace
+
+namespace content_settings {
+
+BravePolicyProvider::BravePolicyProvider(PrefService* prefs)
+    : PolicyProvider(prefs) {
+  ReadManagedDefaultSettings();
+  ReadManagedContentSettings(false);
+  pref_change_registrar_.Init(prefs_);
+  PrefChangeRegistrar::NamedChangeCallback callback = base::BindRepeating(
+      &BravePolicyProvider::OnPreferenceChanged, base::Unretained(this));
+  for (const char* pref : kBraveManagedPrefs)
+    pref_change_registrar_.Add(pref, callback);
+}
+
+BravePolicyProvider::~BravePolicyProvider() = default;
+
+// static
+void BravePolicyProvider::RegisterProfilePrefs(
+    user_prefs::PrefRegistrySyncable* registry) {
+  PolicyProvider::RegisterProfilePrefs(registry);
+  for (const char* pref : kBraveManagedPrefs)
+    registry->RegisterListPref(pref);
+}
+
+void BravePolicyProvider::OnPreferenceChanged(const std::string& name) {
+  if (base::Contains(kBraveManagedPrefs, name)) {
+    ReadManagedContentSettings(true);
+    ReadManagedDefaultSettings();
+  }
+  PolicyProvider::OnPreferenceChanged(name);
+}
+
+void BravePolicyProvider::ReadManagedContentSettings(bool overwrite) {
+  PolicyProvider::ReadManagedContentSettings(overwrite);
+  GetBraveContentSettingsFromPreferences(&value_map_);
+}
+
+/// Duplicating PolicyProvider::GetContentSettingsFromPreferences for
+/// kBravePrefsForManagedContentSettingsMap
+void BravePolicyProvider::GetBraveContentSettingsFromPreferences(
+    OriginIdentifierValueMap* value_map) {
+  for (const auto& entry : kBravePrefsForManagedContentSettingsMap) {
+    // Skip unset policies.
+    if (!prefs_->HasPrefPath(entry.pref_name)) {
+      VLOG(2) << "Skipping unset preference: " << entry.pref_name;
+      continue;
+    }
+
+    const PrefService::Preference* pref =
```

Comment:
all of this duplicated code from `GetContentSettingsFromPreferences` is not ok. Patching code is better than duplicating code so I think we can probably remove this class and just patch `kPrefsForManagedContentSettingsMap` and `kManagedPrefs` (`#define BRAVE_MANAGED_PREFS ...`)

Reply:
done